src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionSource.cs

```diff
@@ -0,0 +1,327 @@
+using VSCompletionItem = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data.CompletionItem;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.AsyncCompletion
+{
+    internal class CompletionSource : ForegroundThreadAffinitizedObject, IAsyncCompletionSource
+    {
+        internal const string RoslynItem = nameof(RoslynItem);
+        internal const string TriggerSnapshot = nameof(TriggerSnapshot);
+        internal const string InsertionText = nameof(InsertionText);
+        internal const string HasSuggestionItemOptions = nameof(HasSuggestionItemOptions);
+        internal const string Description = nameof(Description);
+        internal const string InitialTriggerKind = nameof(InitialTriggerKind);
+        internal const string PotentialCommitCharacters = nameof(PotentialCommitCharacters);
+        internal const string ExcludedCommitCharacters = nameof(ExcludedCommitCharacters);
+        internal const string NonBlockingCompletion = nameof(NonBlockingCompletion);
+
+        private static readonly ImmutableArray<ImageElement> s_WarningImageAttributeImagesArray =
+            ImmutableArray.Create(new ImageElement(Glyph.CompletionWarning.GetImageId(), EditorFeaturesResources.Warning_image_element_automation_name));
+
+        private static readonly EditorOptionKey<bool> NonBlockingCompletionEditorOption = new EditorOptionKey<bool>(NonBlockingCompletion);
+
+        private readonly ITextView _textView;
+
+        internal CompletionSource(ITextView textView, IThreadingContext threadingContext) : base(threadingContext)
+        {
+            _textView = textView;
+        }
+
+        public AsyncCompletionData.CompletionStartData InitializeCompletion(
+            AsyncCompletionData.CompletionTrigger trigger,
+            SnapshotPoint triggerLocation,
+            CancellationToken cancellationToken)
+        {
+            AssertIsForeground();
+
+            var document = triggerLocation.Snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                return AsyncCompletionData.CompletionStartData.DoesNotParticipateInCompletion;
+            }
+
+            var service = document.GetLanguageService<CompletionService>();
+            if (service == null)
+            {
+                return AsyncCompletionData.CompletionStartData.DoesNotParticipateInCompletion;
+            }
+
+            if (!document.Project.Solution.Workspace.Options.GetOption(CompletionOptions.BlockForCompletionItems, service.Language))
+            {
+                _textView.Options.GlobalOptions.SetOptionValue(NonBlockingCompletionEditorOption, true);
+            }
+
+            triggerLocation.Snapshot.TextBuffer.Properties.RemoveProperty(PotentialCommitCharacters);
+            triggerLocation.Snapshot.TextBuffer.Properties.AddProperty(PotentialCommitCharacters, service.GetRules().DefaultCommitCharacters);
```

Comment:
Well, it depends.  I see scenarios where Editor re-uses the session and calls Initialize for the same session again. In case of mixed code, there could be different rules.

@amadeusw to confirm.

---
In reply to: [232848636](https://github.com/dotnet/roslyn/pull/29016#discussion_r232848636) [](ancestors = 232848636)

Reply:
i would def like to understand the lifetimes of all these guys.  in the 'legacy' system, Roslyn controlled all of this itself.  In the new one, it would be good to know if the lifetimes are going to be the same.

Reply:
Actually, checked that this is completionService is defined per project. Therefore, we do not need to update it if exists for specific location.

---
In reply to: [232883310](https://github.com/dotnet/roslyn/pull/29016#discussion_r232883310) [](ancestors = 232883310,232848636)

Reply:
Broker created CompletionSources, then CommitManagers, 
then checks if completion should be created (if we provided applicable to span),
then ItemManagers and then a session attached with all managers and sources. Then, all live until the session dismissed.

---
In reply to: [233701741](https://github.com/dotnet/roslyn/pull/29016#discussion_r233701741) [](ancestors = 233701741,232898625)